java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java

```diff
@@ -0,0 +1,126 @@
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S3510")
+public class HostnameVerifierImplementationCheck extends IssuableSubscriptionVisitor {
+
+  private static final String ISSUE_MESSAGE = "Do not unconditionally return true in this method.";
+
+  private static final TypeCriteria TYPE_CRITERIA_STRING = TypeCriteria.is("java.lang.String");
+  private static final TypeCriteria TYPE_CRITERIA_SSL_SESSION = TypeCriteria.is("javax.net.ssl.SSLSession");
+  private static final MethodMatcher VERIFY_METHOD_MATCHER = MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("javax.net.ssl.HostnameVerifier"))
+    .name("verify")
+    .addParameter(TYPE_CRITERIA_STRING)
+    .addParameter(TYPE_CRITERIA_SSL_SESSION);
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD, Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.kind().equals(Tree.Kind.METHOD)) {
+      checkMethodDefinition((MethodTree) tree);
+    } else if (tree.kind().equals(Tree.Kind.METHOD_INVOCATION)) {
+      ((MethodInvocationTree) tree).arguments().forEach(this::checkForLambdaImplementation);
+    } else {
+      ((NewClassTree) tree).arguments().forEach(this::checkForLambdaImplementation);
+    }
+  }
+
+  private void checkMethodDefinition(MethodTree tree) {
+    BlockTree blockTree = tree.block();
+    if (tree.symbol() != null && VERIFY_METHOD_MATCHER.matches(tree) && blockTree != null) {
+      checkBlock(blockTree);
+    }
+  }
+
+  private void checkForLambdaImplementation(ExpressionTree argument) {
+    if (argument.is(Tree.Kind.LAMBDA_EXPRESSION)) {
+      LambdaExpressionTree lambdaExpressionTree = ((LambdaExpressionTree) argument);
+      Tree lambdaBody = lambdaExpressionTree.body();
+      if (isHostnameVerifierSignature(lambdaExpressionTree)) {
+        if (lambdaBody.is(Tree.Kind.BLOCK)) {
+          checkBlock((BlockTree) lambdaBody);
+        } else if ((lambdaBody.is(Tree.Kind.PARENTHESIZED_EXPRESSION) || lambdaBody.is(Tree.Kind.BOOLEAN_LITERAL))
+          && isTrueLiteral((ExpressionTree) lambdaBody)) {
```

Comment:
Given the implem of isTrueLiteral skipping parenthesis there is no need for the first part of this &&. (in other words : you cannot fully cover this condition). the || should be removed.